Overhaul compaction_manager::task

bhalevy opened the pull request:
The series overhauls the compaction_manager::task design and implementation
by properly layering the functionality between the compaction_manager
that deals with generic task execution, and the per-task business logic that is defined
in a set of classes derived from the generic task class.

While at it, the series introduces `task::state` and a set of helper functions to manage it
to prevent leaks in the statistics, fixing #9974.

Two more stats counter were exposed: `completed_tasks` and a new `postponed_tasks`.

Test: sstable_compaction_test
Dtest: compaction_test.py compaction_additional_test.py

Fixes #9974

raphaelsc reviewed (changes requested):
Looks good to me overall, just a few comments. Very welcome refactoring, Benny. Thanks a lot for working on this

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -314,10 +315,11 @@ class compaction_manager::major_compaction_task : public compaction_manager::tas
+
+        auto prev_sg = co_await coroutine::switch_to(_cm._compaction_controller.sg());
+        co_await t->compact_sstables(std::move(descriptor), _compaction_data);
+        co_await coroutine::switch_to(prev_sg);
```
anyway, I realized that we restore the previous sg once we go out of the coroutine frame, minimizing the impact of a forgotten switch_to() call on previous sg.

bhalevy replied:
> what's preventing us from applying RAII for coroutine's sg switcher?

The problem is that to switch back we need to co_await coroutine::switch_to so it can reschedule the coroutine task with the restored scheduling group, but we can't co_await in the destructor of a RAII object.

bhalevy replied:
Simplified in the next draft version by switch_to covering the whole `run_to` coroutine, including both the compaction_manager::task level logic and the called compaction code.

The original scheduling group is restored automatically when the coroutine co_return:s

bhalevy commented on compaction/compaction.hh:
```diff
@@ -43,6 +43,7 @@ public:
 sstring compaction_name(compaction_type type);
 compaction_type to_compaction_type(sstring type_name);
+std::string_view to_string(compaction_type type);
 
```
right

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -26,13 +26,12 @@ using namespace std::chrono_literals;
 class compacting_sstable_registration {
     compaction_manager* _cm;
-    std::vector<sstables::shared_sstable> _compacting;
+    std::unordered_set<sstables::shared_sstable> _compacting;
```
Sure, will do

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -254,121 +254,123 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(r
+future<> compaction_manager::perform_major_compaction(replica::table* t) {
+    if (_state != state::enabled) {
+        return make_ready_future<>();
+    }
```
I thought about doing that, but then we would create the task in vein.
Maybe if we templatize perform_task, we can let it make the shared task after checking `_state`, basically moving everything in this function to the templated `perform_task`

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -399,25 +401,111 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
+    case state::pending: return "pending";
+    case state::active: return "active";
+    case state::done: return "done";
+    case state::postponed: return "postponed";
```
I used it for auto-incrementing `_cm._stats.postponed_tasks`, but if we change the counter to be stateless, just returning `_postponed.size()` the state won't be needed.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -399,25 +401,111 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
+    if (!_sstables.empty()) {
+        on_internal_error(cmlog, format("sstables were already set"));
+    }
+    _sstables = std::move(new_sstables);
```
> what about creating a class which provides those facilities? then both validate_sstables_compaction_task and rewrite_sstables_compaction_task can inherit from it.

Sounds good.

bhalevy replied:
Fixed in next draft version

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -399,25 +401,111 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
+
 void compaction_manager::task::setup_new_compaction(utils::UUID output_run_id) {
     _compaction_data = create_compaction_data();
     _compaction_running = true;
```
_compaction_running is sticky, where _state can change back and forth between `pending` and `active`, like in `offstrategy_compaction_task::do_run`, so I'm not sure we can get rid of it.

I wonder what it should mean exactly.
It's being used in `compaction_manager::get_compactions()` that is used by:
- `stop_ongoing_compactions` just from printing purposes, and
- by the get_compactions api again for reporting purposes.

Should it count compaction tasks of any kind or only regular compaction?

And what does "ongoing" mean in this context?
That there's an active compaction job running on behalf on the compaction_manager task?
If so, then relying on the `active` state makes sense.

My plan for https://github.com/scylladb/scylla/issues/8392 is to track all compaction_manager tasks once they are created.

bhalevy replied:
Done in next draft version

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -491,6 +491,8 @@ void compaction_manager::register_metrics() {
                        sm::description("Holds the number of currently active compactions.")),
         sm::make_gauge("pending_compactions", [this] { return _stats.pending_tasks; },
                        sm::description("Holds the number of compaction tasks waiting for an opportunity to run.")),
+        sm::make_derive("completed_compactions", [this] { return _stats.completed_tasks; },
```
meaning completed with an error?
we don't currently track that, do we?

raphaelsc replied:
we do. we just need to expose it. Saw you did it already on next version. thanks!

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -493,6 +493,8 @@ void compaction_manager::register_metrics() {
         sm::make_derive("completed_compactions", [this] { return _stats.completed_tasks; },
                        sm::description("Holds the number of completed compaction tasks.")),
+        sm::make_derive("postponed_compactions", [this] { return _stats.postponed_tasks; },
+                       sm::description("Holds the number of postponed compaction tasks.")),
```
okay, makes sense.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -642,329 +733,303 @@ inline bool compaction_manager::can_proceed(replica::table* t) const {
 }
 
-inline bool compaction_manager::task::can_proceed() const {
+inline bool compaction_manager::task::can_proceed(bool throw_if_stopping) const {
```
okay

bhalevy replied:
Done in next draft version

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -930,23 +935,28 @@ class compaction_manager::rewrite_sstables_compaction_task : public compaction_m
+                ex = std::current_exception();
+            }
+            co_await coroutine::switch_to(prev_sg);
+
```
Hmm, good point.  Maybe this is one place we can count failed compactions.

bhalevy replied:
Fixed in next draft version

bhalevy commented on compaction/compaction_manager.hh:
```diff
@@ -112,6 +116,12 @@ private:
+    struct regular_compaction_task;
+    struct offstrategy_compaction_task;
+    struct rewrite_sstables_compaction_task;
+
```
I don't remember exactly, I'll see if they can be removed after `class task` is made public in compaction_manager, so it can be inherited from by classes unrelated to compaction_manager`.

bhalevy replied:
It turns out that the tasks do need access to private compaction_manager members.
(E.g. `_maintenance_ops_sem`, `postponed_compactions_reevaluation()`, `reevaluate_postponed_compactions()`)

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -255,7 +255,7 @@ future<> compaction_manager::perform_major_compaction(replica::table* t) {
 
-    auto task = make_lw_shared<compaction_manager::task>(t, sstables::compaction_type::Compaction, get_compaction_state(t));
+    auto task = make_shared<compaction_manager::task>(t, sstables::compaction_type::Compaction, get_compaction_state(t));
     _tasks.push_back(task);
```
I think that the issue is with `shared_ptr_no_esft` having `T _value` as a member.
Therefore `T` has to be fully defined.

(That said, the struct name is misleading. It should be called `lw_shared_ptr_no_esft`)

In contrast `shared_ptr` only has a `T* _p` member, and so it's satisfied with a forward declaration of T.

raphaelsc reviewed (commented):
please push the branch again without the first 4 patches sent to mailing list

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -399,25 +401,116 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
+    _sstables = std::move(new_sstables);
+    cmlog.debug("{}: set_sstables: {} sstable{}", *this, _sstables.size(), _sstables.size() > 1 ? "s" : "");
+    _cm._stats.pending_tasks += _sstables.size() - (_state == state::pending);
+}
```
as we unconditionally remove _sstables.size() on task dtor, I think it's better to just increment pending_tasks by _sstables.size(). I know that once we bump into semaphores, pending_tasks will be off-by-one, but that's ok I think. pending tasks is hard to reason about as it doesn't say anything about the weight of individual operations.

bhalevy replied:
> as we unconditionally remove _sstables.size()

This is no longer the case.
See `sstables_task::~sstables_task()`

raphaelsc reviewed (changes requested):
Looks good to me overall. Just one nitpick left.

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -254,121 +254,123 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(r
         return make_ready_future<>();
     }
+    return perform_task(make_shared<custom_compaction_task>(*this, t, type, desc, std::move(job)));
+}
```
@bhalevy I think we should merge run_custom_task_job() into run_custom_job(). It's better not to expose compaction_manager::task in any public interface, and it's also not needed anywhere. Let's only leave run_custom_job().

bhalevy replied:
fixed

raphaelsc commented on compaction/compaction_manager.cc:
```diff
@@ -399,25 +401,116 @@ compaction_manager::run_with_compaction_disabled(replica::table* t, std::functio
+        ++_cm._stats.completed_tasks;
+        break;
+    }
+    cmlog.debug("{}: switch_state: {} -> {}: pending={} active={} done={} errors={}", *this, _state, new_state,
```
_state refers to new_state, after std::exchange(...) above, so we have to use old_state instead for the old state.

bhalevy replied:
True.  I moved the printout from the beginning of the function to the end, and missed that.  Will fix.

raphaelsc reviewed (approved):
Reviewed-with-a-smile-by: Raphael S. Carvalho

bhalevy commented:
@scylladb/scylla-maint please merge.

bhalevy commented:
Converted back to draft, since I think we need to fix #10175 first so the fix could be backported more easily.

avikivity commented on compaction/compaction_manager.cc:
```diff
@@ -293,6 +294,8 @@ class compaction_manager::major_compaction_task : public compaction_manager::tas
     // prevented while an ongoing major compaction doesn't release the semaphore.
     virtual future<> do_run() override {
+        co_await coroutine::switch_to(_cm._compaction_controller.sg());
+
```
What about switching back?

Maybe all switching should be done by common code.

bhalevy replied:
The original scheduling group is restored when the coroutine completes.
See scylladb/seastar@6d061b7adaa482e38744c6a78790d71e482d82be

avikivity commented:
Looks good.

denesb commented on compaction/compaction.hh:
```diff
@@ -43,6 +43,7 @@ public:
 
 sstring compaction_name(compaction_type type);
 compaction_type to_compaction_type(sstring type_name);
+std::string_view to_string(compaction_type type);
```
This needs a comment to explain the difference v.s. `compaction_name()`.

bhalevy replied:
np

bhalevy replied:
added documentation in next version.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -119,6 +163,13 @@ public:
+    class regular_compaction_task;
+    class offstrategy_compaction_task;
+    class rewrite_sstables_compaction_task;
+    class compaction_manager_test_task;
```
Any reason these have to be exposed here?

bhalevy replied:
Unfortunately they need access to compaction_manager private members.

bhalevy replied:
We should cleanup the api the compaction_manager exposes for the tasks, but let's do it as a follow-up.

denesb replied:
Ok, follow-up is fine.

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -72,44 +72,88 @@ private:
+    class task {
+    protected:
+        compaction_manager& _cm;
+        replica::table* _compacting_table = nullptr;
```
If this is never null in practice, it should be a reference (can be a follow-up).

bhalevy replied:
We use a pointer for historical reasons that required loose coupling.
I think that changing it to a reference is possible, but it's a bit out of scope for this series.
Let's consider doing it as a followup.

bhalevy replied:
Will defer for a followup patch/series.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -254,118 +254,118 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(r
-    });
-    return task->compaction_done().then([task] {});
+        finish_compaction();
+    }
```
Coroutinizing should've been a separate patch.

denesb commented on compaction/compaction_manager.cc:
```diff
@@ -639,327 +732,306 @@ inline bool compaction_manager::can_proceed(replica::table* t) const {
+                co_return;
+            }
+        }
+    }
```
ditto

denesb reviewed (approved):
LGTM, nice cleanup!
Nits only.

[compaction_manager: move per-type logic to derived task](https://github.com/scylladb/scylla/pull/10122/commits/d7a727983f914ab0aa133000672d949a0983f3b7) was really hard to review, that patch does so many things. A separate patch per compaction type would have been more digestible (with coroutinization deferred to yet another patch). But as the reviewing is mostly done, I won't insist on splitting it.

bhalevy commented on compaction/compaction_manager.cc:
```diff
@@ -254,118 +254,118 @@ compaction_manager::compaction_state& compaction_manager::get_compaction_state(r
     }
```
will do

bhalevy replied:
done in next version

bhalevy commented:
- Cleaned up the "compaction_manager: move per-type logic to derived task" patch a bit
  - splitting independent patches out of it
  - per-type transition and coroutinization proved to be too hard and time consuming, especially using the new task::state properly that I couldn't guarantee bisectability - so I dropped the effort and stayed with the enormous, yet correct patch (I believe)
  - One fix that was needed was to restrict the exponential retry reset logic in finish_compaction to the success case, since it was called before maybe_retry and if compaction failed and retry was desired, resetting the exponential backoff shouldn't be done.
- Added some commentary for compaction_type stringifying functions
- And documentation for task::state

denesb commented on compaction/compaction_manager.hh:
```diff
@@ -75,11 +75,28 @@ private:
+                        // counted in compaction_manager::stats::pending_tasks
+            active,     // task initiated active compaction, may alternate with pending
+                        // counted in compaction_manager::stats::active_tasks
+            done,       // task compelted successfully (may transition only to state::none)
```
s/compelted/completed/

bhalevy replied:
damn, will fix

denesb reviewed (approved):
LGTM, please fix that one typo and I'll queue.

bhalevy commented:
@denesb I force pushed the branch (also made a minor correction to the comment)

```diff
$ git diff 32d9069 5e1fda7
diff --git a/compaction/compaction_manager.hh b/compaction/compaction_manager.hh
index eec0a3c6d..88ccb7223 100644
--- a/compaction/compaction_manager.hh
+++ b/compaction/compaction_manager.hh
@@ -82,10 +82,10 @@ class compaction_manager {
                         // counted in compaction_manager::stats::pending_tasks
             active,     // task initiated active compaction, may alternate with pending
                         // counted in compaction_manager::stats::active_tasks
-            done,       // task compelted successfully (may transition only to state::none)
+            done,       // task completed successfully (may transition only to state::none)
                         // counted in compaction_manager::stats::completed_tasks
             postponed,  // task was postponed (may transition only to state::none)
-                        // counted in compaction_manager::stats::postponed_tasks
+                        // represented by the postponed_compactions metric
             failed,     // task failed (may transition only to state::none)
                         // counted in compaction_manager::stats::errors
         };
```

denesb commented:
Thanks, queued.
